File: src/containers/hl-engine/tests/fixtures/sentry_process_table_lifetime.c
// Guest programs for the sentry's per-process virtual descriptor table lifetime.
//
// Under `HL_UNTRUSTED` the sentry keys one table per worker PROCESS, by the HOST pid every request from
// that worker carries. The table must live exactly as long as the process: an entry that outlives its
// process holds that process's duplicated descriptors forever, consumes one of the sentry's bounded
// process slots, and -- because the host kernel reissues a freed pid -- can be found by a completely
// different process that is later handed the same number.
//
// argv[1] selects the program:
//   rounds-waitid     one child at a time, ended by a signal and collected with waitid(2)
//   bound             the simultaneously-live bound: fork until the sentry refuses
//   collide <dir>     leak one entry, then fork again onto that same host pid (see the harness)
#define _GNU_SOURCE
#include <errno.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/wait.h>
#include <unistd.h>

#define ROUNDS 200
#define ATTEMPTS 70

static void arm_nocldwait(void) {
    struct sigaction action;
    memset(&action, 0, sizeof action);
    action.sa_handler = SIG_DFL;
    action.sa_flags = SA_NOCLDWAIT;
    if (sigaction(SIGCHLD, &action, NULL) != 0) exit(20);
}

// One child alive at a time, for ROUNDS rounds. Every round ends its child with a signal, so the child
// never publishes its own exit, and collects it by a route other than wait4(2). A sentry that releases the
// entry when the host pid is freed runs this to completion; one that does not exhausts its slots instead.
static int rounds(void) {
    int completed = 0, failure = 0;
    for (int i = 0; i < ROUNDS; i++) {
        pid_t child = fork();
        if (child == 0)
            for (;;) pause();
        if (child < 0) { failure = errno; break; }
        kill(child, SIGKILL);
        siginfo_t info;
        memset(&info, 0, sizeof info);
        while (waitid(P_PID, (id_t)child, &info, WEXITED) < 0 && errno == EINTR) {}
        completed++;
    }
    printf("rounds completed=%d failure=%d\n", completed, failure);
    return completed == ROUNDS ? 0 : 1;
}

// The bound on tables the sentry holds for SIMULTANEOUSLY live children. This is fail-closed behaviour and
// must not move: a fork past it is refused, not served from a recycled slot.
static int bound(void) {
    pid_t children[ATTEMPTS];
    int created = 0, failure = 0;
    for (int i = 0; i < ATTEMPTS; i++) {
        pid_t child = fork();
        if (child == 0)
            for (;;) pause();
        if (child < 0) { failure = errno; break; }
        children[created++] = child;
    }
    int reaped = 0;
    for (int i = 0; i < created; i++) {
        kill(children[i], SIGKILL);
        int status;
        if (waitpid(children[i], &status, 0) == children[i]) reaped++;
    }
    printf("bound created=%d failure=%d reaped=%d\n", created, failure, reaped);
    return created == 63 && failure == EAGAIN && reaped == created ? 0 : 1;
}

static void publish(const char *path, const char *text) {
    FILE *file = fopen(path, "w");
    if (!file) exit(21);
    fputs(text, file);
    fclose(file);
}

static int await(const char *path, int seconds) {
    for (int i = 0; i < seconds * 1000; i++) {
        FILE *file = fopen(path, "r");
        if (file) { fclose(file); return 1; }
        usleep(1000);
    }
    return 0;
}

// Leak exactly one entry, then ask the harness to re-arm the kernel's pid allocator on the leaked number
// and fork again. The fork must be SERVED: the pid names a child this process was just handed, so any
// entry filed under it is a corpse. Refusing it (-EEXIST, an errno clone(2) cannot return) fails an
// ordinary container fork on a number the kernel was free to reissue.
//
// The leak is arranged with SA_NOCLDWAIT because that is the release path the sentry still does not see;
// if that path is ever closed this program no longer leaks, the fork is served for the plainer reason,
// and the assertion below stays true.
static int collide(const char *directory) {
    char victim_path[512], go_path[512], text[128];
    snprintf(victim_path, sizeof victim_path, "%s/victim", directory);
    snprintf(go_path, sizeof go_path, "%s/go", directory);
    arm_nocldwait();
    pid_t victim = fork();
    if (victim == 0)
        for (;;) pause();
    if (victim < 0) return 2;
    kill(victim, SIGKILL);
    usleep(20000);
    for (int attempt = 1; attempt <= 60; attempt++) {
        snprintf(text, sizeof text, "%d %d\n", (int)victim, attempt);
        publish(victim_path, text);
        if (!await(go_path, 30)) return 3;
        pid_t child = fork();
        int error = errno;
        if (child == 0) _exit(0);
        remove(go_path);
        if (child < 0) {
            printf("collide refused attempt=%d errno=%d\n", attempt, error);
            return 7;
        }
        if (child == victim) {
            printf("collide served victim=%d attempt=%d\n", (int)victim, attempt);
            return 0;
        }
    }
    printf("collide no-collision\n");
    return 6;
}

int main(int argc, char **argv) {
    const char *program = argc > 1 ? argv[1] : "";
    if (strcmp(program, "rounds-waitid") == 0) return rounds();
    if (strcmp(program, "bound") == 0) return bound();
    if (strcmp(program, "collide") == 0 && argc > 2) return collide(argv[2]);
    return 30;
}

File: src/containers/hl-engine/tests/sentry_process_table_lifetime.rs
#![cfg(target_os = "linux")]
//! The lifetime of the sentry's per-process virtual descriptor table, under the production sandbox default.
//!
//! `Sandbox::SentryOnly` sets `HL_UNTRUSTED`, and the sentry then keys one table per worker PROCESS by the
//! HOST pid every request from that worker carries. Two properties have to hold together and neither is
//! visible without a guest process tree:
//!
//! * the table is released when its process dies, whichever route collects the corpse -- otherwise it holds
//!   that process's duplicated descriptors forever and occupies one of the sentry's bounded process slots;
//! * a host pid the kernel has reissued names a different process, and the fork lane must serve it. Measured
//!   on this `x86_64` box: the kernel reissued a specific freed pid after 40.9 s of ordinary fork churn, so
//!   this is a matter of container uptime, not of probability.

use hl_engine::{activation::GuestIsa, launcher::plan::RuntimePlan, options::Options, runtime::Engine};
use std::path::{Path, PathBuf};
use tempfile::TempDir;

fn fixture(directory: &Path) -> PathBuf {
    let source = Path::new(env!("CARGO_MANIFEST_DIR")).join("tests/fixtures/sentry_process_table_lifetime.c");
    let output = directory.join("sentry-process-table-lifetime");
    let compiler = "x86_64-linux-gnu-gcc";
    let status = std::process::Command::new(compiler)
        .args(["-static", "-O2", "-std=gnu11", "-o"])
        .arg(&output)
        .arg(source)
        .status()
        .unwrap_or_else(|error| panic!("cannot run {compiler}: {error}"));
    assert!(status.success(), "{compiler} failed with {status}");
    output
}

/// Runs the fixture under the sandbox default and answers its guest exit status.
fn sandboxed(executable: &Path, arguments: &[&str]) -> i32 {
    let mut options = Options::default();
    options.set("HL_UNTRUSTED", "1", true).unwrap();
    let mut argv = vec![executable.as_os_str().as_encoded_bytes().to_vec()];
    argv.extend(arguments.iter().map(|argument| argument.as_bytes().to_vec()));
    let plan = RuntimePlan {
        rootfs: None,
        executable_host: Some(executable.as_os_str().as_encoded_bytes().to_vec()),
        arguments: argv,
        environment: Vec::new(),
        result_path: None,
        options,
    };
    let engine = Engine::from_plan(GuestIsa::X86_64, plan).expect("sandboxed launch");
    engine.start().expect("guest start");
    engine.wait().expect("guest wait").guest_status
}

/// `wait4(2)` had a sentry lane and `waitid(2)` did not, so a guest that ends a child with a signal -- the
/// child then never publishes its own exit -- and collects it with `waitid` freed the host pid while
/// leaving the table filed under it. One child is alive at a time here, so the only thing 200 rounds can
/// exhaust is entries that outlived their processes; before the lane existed this stopped at 63 with
/// `EAGAIN`.
#[test]
fn a_child_collected_with_waitid_releases_its_descriptor_table() {
    let work = TempDir::new().unwrap();
    let executable = fixture(work.path());
    assert_eq!(
        sandboxed(&executable, &["rounds-waitid"]),
        0,
        "a signalled child collected with waitid left its descriptor table behind"
    );
}

/// The bound on tables held for SIMULTANEOUSLY live children is fail-closed and must stay where it is: a
/// fork past it is refused with `EAGAIN`, never served out of a slot something else still owns. This is the
/// control for the test above -- without it, "200 rounds completed" would also be satisfied by a sentry
/// that had stopped bounding anything.
#[test]
fn the_bound_on_simultaneously_live_children_is_unchanged() {
    let work = TempDir::new().unwrap();
    let executable = fixture(work.path());
    assert_eq!(
        sandboxed(&executable, &["bound"]),
        0,
        "the sentry no longer refuses the 64th simultaneously live child with EAGAIN"
    );
}

/// A host pid the kernel reissues names a new process, and the fork lane holds the proof: `clone(2)` just
/// handed this worker that number, so no other live process can own it and any entry filed under it is a
/// corpse. The lane must reclaim it and serve the fork. Refusing instead returned `-EEXIST`, which
/// `clone(2)` cannot return on Linux, from an ordinary container fork.
///
/// Rather than churn 4.2 million pids, the harness re-arms `ns_last_pid` on the number the guest leaked,
/// once per guest attempt. That needs privilege, and every statement in the arming loop has to avoid
/// spawning a process -- one intervening fork consumes the pid being armed.
#[test]
fn a_fork_onto_a_reissued_host_pid_is_served() {
    let Ok(last_pid) = std::fs::OpenOptions::new()
        .write(true)
        .open("/proc/sys/kernel/ns_last_pid")
    else {
        println!(
            "SKIP a_fork_onto_a_reissued_host_pid_is_served: /proc/sys/kernel/ns_last_pid is not writable \
             here, so a reissued pid cannot be arranged without churning the whole pid space"
        );
        return;
    };
    drop(last_pid);
    let work = TempDir::new().unwrap();
    let executable = fixture(work.path());
    let handshake = work.path().to_path_buf();
    let arming = std::thread::spawn(move || {
        let victim = handshake.join("victim");
        let go = handshake.join("go");
        let stop = handshake.join("stop");
        let mut armed = String::new();
        while !stop.exists() {
            let Ok(request) = std::fs::read_to_string(&victim) else {
                continue;
            };
            if request.trim().is_empty() || request == armed {
                continue;
            }
            let Some(number) = request.split_whitespace().next().and_then(|f| f.parse::<i32>().ok()) else {
                continue;
            };
            armed = request;
            let _ = std::fs::write("/proc/sys/kernel/ns_last_pid", format!("{}", number - 1));
            let _ = std::fs::write(&go, []);
        }
    });
    let status = sandboxed(&executable, &["collide", &work.path().to_string_lossy()]);
    std::fs::write(work.path().join("stop"), []).unwrap();
    arming.join().unwrap();
    assert_ne!(
        status, 7,
        "a fork onto a reissued host pid was refused rather than served"
    );
    assert!(
        status == 0 || status == 6,
        "the collision fixture failed for its own reasons: guest status {status}"
    );
}
